[clang][modules] Guard against bad -fmodule-file mappings (#132059)

naveen-seth opened the pull request:
Fix https://github.com/llvm/llvm-project/issues/132059.

Providing incorrect mappings via `-fmodule-file=<name>=<path/to/bmi>`
can crash the compiler when loading a module that imports an
incorrectly mapped module.

The crash occurs during AST body deserialization, when the compiler
attempts to resolve remappings using the `ModuleFile` from the
incorrectly mapped module's BMI file.
The cause is an invalid access into an incorrectly loaded
`ModuleFile`.

This commit fixes the issue by verifying the identity of the imported
module.

ChuanqiXu9 reviewed (commented):
The size of the patch is slightly more than what I thought. I thought this may be a simple change. Could you describe the crash point and the crash reason (from code's perspective)

naveen-seth commented:
At first I thought the issue could be fixed by just checking for duplicates
while parsing the command-line arguments.

I will try to explain the crash point and reason with an example:

```cpp
//--- a.cppm
export module A;

export int a() {
  return 41;
}
```

```cpp
//--- b.cppm
export module B;
import A;

export int b() {
  return a() + 1;
}
```

```cpp
//--- main.cpp
import A;

int main() {
  return a();
}
```

```bash
# Correctly precompile the modules
clang++ -std=c++20 --precompile a.cppm -o a.pcm
clang++ -std=c++20 --precompile b.cppm -fmodule-file=A=a.pcm -o b.pcm
# Bad use of -fmodule-file=...
clang++ -std=c++20 main.cpp -fmodule-file=A=a.pcm -fmodule-file=A=b.pcm
```


Looking at how the arguments are parsed [(here)](https://github.com/llvm/llvm-project/blob/191e0622e84a215c1d632412561c2fe6ccd86170/clang/lib/Frontend/CompilerInvocation.cpp#L3388C16-L3388C23), the mapping `A=a.pcm` is overridden by `A=b.pcm`.
Therefore, it is no different than:

```bash
clang++ -std=c++20 main.cpp -fmodule-file=A=b.pcm
```

The mismatch here can't be detected by validating the command-line arguments.
Clang can only detect the mismatch once it has read the module name in `b.pcm`.

Let's now look at how the error occurs:

(from top to bottom)

`Sema::ActOnModuleImport`
`CompilerInstance::loadModule`

`b.pcm` is found as the BMI to load the module from.

`CompilerInstance::findOrCompileModuleAndReadAST`
`ASTReader::ReadAST`
`ASTReader::ReadASTCore`

Module `A` is added to `ModuleMgr` and the AST for `b.pcm` is read because `ModuleMgr` reports it as newly loaded.

`ASTReader::ReadControlBlock`

Currently (prior to this PR), there is no validation to ensure the BMI file actually contains the module it is trying to load after having read the module name in `ASTReader::ReadControlBlock`.
The control block continues to be read.
When the import of `A` in the control block of `b.pcm` is reached, clang
attempts to load again.
The `ModuleMgr` will report the AST of `b.pcm` as already loaded and no error is generated at this point either.

We now effectively have a recursive import.

The crash then occurs when reading the AST blocks and trying to resolve the module file remappings (DeclID, SubmoduleID, ...), because of invalid access into members of the invalid `ModuleFile` for module `A`.

ChuanqiXu9 commented:
I received a mail but not see it in the web page. Is something going wrong.

naveen-seth commented:
Hi, sorry! I left a comment earlier but noticed that a large part of the explanation was incorrect and deleted it. Here is the fixed comment:

At first I thought that the issue can be fixed by just checking for duplicates while parsing the command-line arguments, but that doesn't seem to be the case.

I will try to describe the crash point and the crash reason with this example (without duplicate arguments):

```cpp
//--- a.cppm
export module A;

export int a() {
  return 41;
}
```

```cpp
//--- b.cppm
export module B;
import A;

export int b() {
  return a() + 1;
}
```

```cpp
//--- c.cppm
export module C;

export int c() {
  return 42;
}
```

```cpp
//--- main.cpp
import B;

int main() {
  return b();
}
```

```
# Correctly precompile the modules
clang++ -std=c++20 --precompile a.cppm -o a.pcm
clang++ -std=c++20 --precompile b.cppm -fmodule-file=A=a.pcm -o b.pcm
clang++ -std=c++20 --precompile c.cppm -o c.pcm

# Bad use of -fmodule-file=...
clang++ -std=c++20 main.cpp -std=c++20 main.cpp -fmodule-file=B=b.pcm -fmodule-file=A=c.pcm
```

After `import B;` is parsed from `main.cpp`, the following sequence of function calls occurs (from top to bottom): 

…
`Sema::ActOnModuleImport`
`CompilerInstance::loadModule`
`CompilerInstance::findOrCompileModuleAndReadAST`
`ASTReader::ReadAST`
`ASTReader::ReadASTCore`
`ASTReader::ReadControlBlock`

When reading a modules control block in `ASTReader::ReadControlBlock`, Clang also attempts to get the `ModuleFile` for each of its imports ([see here](https://github.com/llvm/llvm-project/blob/943a70717c629f43b309ab56e8141ffb131871a6/clang/lib/Serialization/ASTReader.cpp#L3024)).
First, the matching BMI file for an imported module is looked up, then Clang calls back to `ASTReader::ReadASTCore` to ensure the BMI files `ModuleFile`.

Currently (prior to this PR), there is no validation to ensure that the BMI file actually contains the module clang is trying to load after having read the module name in `ReadControlBlock`.

Because of the faulty `-fmodule-file=` mapping, the `ModuleFile` of the import will refer to a BMI built for a different module than the one clang is trying to load.
Here, the import of `A` in `b.cppm` will result in loading the `ModuleFile` corresponding to `c.pcm` while deserializing `b.pcm`. 

The crash occurs because Clang, assuming it has the correct `ModuleFile`s, continues reading AST blocks and attempts to resolve `ModuleFile` remappings (any of the mappings mentioned [here](https://github.com/llvm/llvm-project/blob/74ace4fd3c71ff59f2d89680c0f1382d0f9933f4/clang/lib/Serialization/ASTReader.cpp#L8765)). This results in invalid access into members of the invalid `ModuleFile`s and finally the crash.

(Depending on the specific example, the exact location of the crash will be different, but it will always happen while trying to resolve the remappings.)

While writing this comment, I noticed that my commit and PR message describe the cause of the crash incorrectly and have edited both.

ChuanqiXu9 reviewed (commented):
Thanks. What I feel not good is passing `ExpectedName` and `ModuleMismatch` all around. I am wondering if we can check it by inserting a check in `ASTReader::ReadControlBlock` in `case IMPORT:` where we can check the `ImportedName` and the name stored in loaded module.

naveen-seth committed:
[clang][modules] Guard against bad -fmodule-file mappings (#132059)

[clang][modules] Guard against invalid -fmodule-file mappings

Fixes #132059.

Providing incorrect mappings via `-fmodule-file=<name>=<path/to/bmi>`
can cause the compiler to crash when loading a module that imports an
incorrectly mapped module.

The crash occurs during AST body deserialization when the compiler
attempts to resolve remappings using the `ModuleFile` from the
incorrectly mapped module's BMI file.
The cause is an invalid access into an incorrectly loaded
`ModuleFile`.

This commit fixes the issue by verifying the identity of the imported
module.

naveen-seth commented:
Updated! That really makes does make it a lot simpler.

I was a bit unsure regarding the following lines:

https://github.com/llvm/llvm-project/blob/b98d66f394cc2e25ea0873cfda79704e10795c7b/clang/lib/Serialization/ASTReader.cpp#L3342-L3346 

Is it acceptable to also return `OutOfDate` for this error case?
(We don´t return immediately to also include the backtrace.)

Thank you for the review and my apologies for the delay!

ChuanqiXu9 reviewed (approved):
LGTM

ChuanqiXu9 commented on clang/lib/Serialization/ASTReader.cpp:
```diff
@@ -3318,6 +3318,14 @@ ASTReader::ReadControlBlock(ModuleFile &F,
+      // Check the AST we just read from ImportedFile contains a different
+      // module than we expected (ImportedName).
+      if (auto *Imported = getModuleManager().lookupByFileName(ImportedFile);
+          Imported != nullptr && Imported->ModuleName != ImportedName) {
```
To make it more controllable, let's add a check for `IsImportingStdCXXModule`

naveen-seth replied:
Added.
Theoretically also only applies to `MK_PrebuiltModule`. 
Should I also add it to the check?

ChuanqiXu9 reviewed (approved):
LGTM. Thanks.
